download/oci_dowload.go

```diff
@@ -0,0 +1,379 @@
+func (d *OCIDownloader) SetCache(etag string) {
+}
+
+// Trigger can be used to control when the downloader attempts to download
+// a new bundle in manual triggering mode.
+func (d *OCIDownloader) Trigger(ctx context.Context) error {
+	done := make(chan error)
+
+	go func() {
+		err := d.oneShot(ctx)
+		if err != nil {
+			d.logger.Error("OCI - Bundle download failed: %v.", err)
+			if ctx.Err() == nil {
+				done <- err
+			}
+		}
+		close(done)
+	}()
+
+	select {
+	case err := <-done:
+		return err
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
+
+// Start tells the Downloader to begin downloading bundles.
+func (d *OCIDownloader) Start(ctx context.Context) {
+	if *d.config.Trigger == plugins.TriggerPeriodic {
+		go d.doStart(ctx)
+	}
+}
+
+// Stop tells the Downloader to stop downloading bundles.
+func (d *OCIDownloader) Stop(context.Context) {
+	if *d.config.Trigger == plugins.TriggerManual {
+		return
+	}
+
+	d.mtx.Lock()
+	defer d.mtx.Unlock()
+
+	if d.stopped {
+		return
+	}
+
+	done := make(chan struct{})
+	d.stop <- done
+	<-done
+}
+
+func (d *OCIDownloader) doStart(context.Context) {
+	// We'll revisit context passing/usage later.
+	ctx, cancel := context.WithCancel(context.Background())
+
+	d.wg.Add(1)
+	go d.loop(ctx)
+
+	done := <-d.stop // blocks until there's something to read
+	cancel()
+	d.wg.Wait()
+	d.stopped = true
+	close(done)
+}
+
+func (d *OCIDownloader) loop(ctx context.Context) {
+	defer d.wg.Done()
+
+	var retry int
+
+	for {
+
+		var delay time.Duration
+
+		err := d.oneShot(ctx)
+
+		if ctx.Err() != nil {
+			return
+		}
+
+		if err != nil {
+			delay = util.DefaultBackoff(float64(minRetryDelay), float64(*d.config.Polling.MaxDelaySeconds), retry)
+		} else {
+			if !d.longPollingEnabled || d.config.Polling.LongPollingTimeoutSeconds == nil {
+				// revert the response header timeout value on the http client's transport
+				min := float64(*d.config.Polling.MinDelaySeconds)
+				max := float64(*d.config.Polling.MaxDelaySeconds)
+				delay = time.Duration(((max - min) * rand.Float64()) + min)
+			}
+		}
+
+		d.logger.Debug("OCI - Waiting %v before next download/retry.", delay)
+
+		select {
+		case <-time.After(delay):
+			if err != nil {
+				retry++
+			} else {
+				retry = 0
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (d *OCIDownloader) oneShot(ctx context.Context) error {
+	m := metrics.New()
+	resp, err := d.download(ctx, m)
+	if err != nil {
+		if d.f != nil {
+			d.f(ctx, Update{ETag: "", Bundle: nil, Error: err, Metrics: m, Raw: nil})
+		}
+		return err
+	}
+
+	d.longPollingEnabled = resp.longPoll
+
+	if d.f != nil {
+		d.f(ctx, Update{ETag: resp.etag, Bundle: resp.b, Error: nil, Metrics: m, Raw: resp.raw})
+	}
+	return nil
+}
+
+func (d *OCIDownloader) download(ctx context.Context, m metrics.Metrics) (*downloaderResponse, error) {
+	d.logger.Debug("OCI - Download starting.")
+
+	preferences := []string{fmt.Sprintf("modes=%v,%v", defaultBundleMode, deltaBundleMode)}
```

Comment:
Can delta bundles be used as well ? I guess as long as we have a tarball it should be supported , correct ?

Reply:
As @viovanov stated here https://github.com/open-policy-agent/opa/issues/4518#issuecomment-1102261438 this should work